src/door.rs

```diff
@@ -0,0 +1,291 @@
+                    eprintln!("{0}", e);
+                    std::process::exit(1);
+                }
+            }
+        }
+        None => user,
+    };
+}
+
+fn get_container_id(child: Child) -> String {
+    return match child.wait_with_output() {
+        Ok(output) => match output.status.code() {
+            Some(code) => {
+                if code != 0 {
+                    eprintln!("Starting container failed with status {0}", code);
+                    std::process::exit(1);
+                }
+                match String::from_utf8(output.stdout) {
+                    Ok(container_id) => container_id,
+                    Err(e) => {
+                        eprintln!("Couldn't decode container ID: {0}", e);
+                        std::process::exit(1);
+                    }
+                }
+            }
+            None => {
+                eprintln!("Starting container failed with unknown status");
+                std::process::exit(1);
+            }
+        },
+        Err(e) => {
+            eprintln!("Starting container failed: {0}", e);
+            std::process::exit(1);
+        }
+    };
+}
+
+pub fn launch(args: &LaunchArgs, config: &Config) {
+    let door = get_door(&args.door, config);
+    let user = switch_user(get_caller(config), config, &args.user);
+
+    let rundir = Path::new(&config.doorman.rundir);
+
+    match fs::create_dir_all(rundir) {
+        Ok(_) => (),
+        Err(e) => {
+            eprintln!("Couldn't create rundir {0}: {1}", rundir.display(), e);
+            std::process::exit(1);
+        }
+    };
+
+    let door_lockfile_path = rundir.join(format!("{0}.lock", args.door));
+    let door_lockfile = match fs::File::options()
+        .read(true)
+        .write(true)
+        .create(true)
+        .truncate(false)
+        .open(&door_lockfile_path)
+    {
+        Ok(lockfile) => lockfile,
+        Err(e) => {
+            eprintln!(
+                "Couldn't open lockfile {0}: {1}",
+                door_lockfile_path.display(),
+                e
+            );
+            std::process::exit(1);
+        }
+    };
+
+    match door_lockfile.try_lock_shared() {
+        Ok(_) => (),
+        Err(_) => {
+            eprintln!("Sorry, {0} is currently undergoing maintenence.", args.door);
+            std::process::exit(1);
+        }
+    };
+
+    let mut node: i8 = 1;
+    let mut found_node = false;
+
+    while (node <= door.nodes) && !found_node {
+        let node_lockfile_path = rundir.join(format!("{0}.{1}.lock", args.door, node));
+        let node_lockfile = fs::File::options()
+            .read(true)
+            .write(true)
+            .create(true)
+            .truncate(false)
+            .open(&node_lockfile_path);
+
+        if node_lockfile.is_ok() {
+            found_node = match node_lockfile.unwrap().try_lock_exclusive() {
+                Ok(_) => true,
+                Err(_) => false,
+            };
+        }
+
+        if found_node {
```

Comment:
I feel like this should be a guard clause (`if !found_node { continue; }` followed by the rest, de-indented). Or even something like:

```rs
if !node_lockfile.and_then(|lf| lf.try_lock_exclusive().ok()).is_some() {
  continue;
}

found_node = true;
// the rest of the code
```

(not sure how exactly `found_node` is used elsewhere, btw, but *setting* it to `true`/`false` (vs setting it to `true` and not resetting it if `false`) might be a bug? just based on the last bit where it's checked to see if *all* nodes are busy, when it actually only represents the *last* node.)

Reply:
`found_node` should never be set back to `false` once it's `true`; the intention is to increment `node` until we've found one we can lock, run the door, and break out of the loop. If it's still `false` at the end of the loop, that means we went through all the available nodes w/o finding one that can be locked.

Reply:
Then your current code has a bug, yep, as you occasionally set it back to `false` (if the lockfile was opened but locking it fails).

Reply:
The loop shouldn't execute again once it's set to `true`? `!found_node` is part of the loop condition.

Reply:
I've restructured it